node/src/components/small_network/tasks.rs

```diff
@@ -353,6 +362,34 @@ where
             return Err(ConnectionError::WrongNetwork(network_name));
         }
 
+        // If there is a version mismatch, we treat it as a connection error. We do not ban peers
+        // for this error, but instead rely on exponential backoff, as bans would result in issues
+        // during upgrades where nodes may have a legitimate reason for differing versions.
+        //
+        // Since we are not using SemVer for versioning, we cannot make any assumptions about
+        // compatibility, so we allow only exact version matches.
+        if context.reject_incompatible_versions
+            && protocol_version != context.chain_info.protocol_version
+        {
+            if let Some(threshold) = context.tarpit_version_threshold {
+                if protocol_version <= threshold {
+                    let mut rng = OsRng;
+
+                    let sample = rng.gen_range(0.0f32..1.0f32);
+                    if context.tarpit_chance > sample {
```

Comment:
We could use [`gen_bool`](https://rust-random.github.io/rand/rand/trait.Rng.html#method.gen_bool):
```suggestion
                    if rng.gen_bool(context.tarpit_chance as f64) {
```

Reply:
TIL! Thanks.